helix-lsp/src/lib.rs

```diff
@@ -263,83 +307,140 @@ pub mod util {
         snippet: snippet::Snippet,
         line_ending: &str,
         include_placeholder: bool,
+        tab_width: usize,
     ) -> Transaction {
         let text = doc.slice(..);
+        let mut new_selection: SmallVec<[_; 1]> = SmallVec::new();
 
-        // For each cursor store offsets for the first tabstop
-        let mut cursor_tabstop_offsets = Vec::<SmallVec<[(i128, i128); 1]>>::new();
-        let transaction = Transaction::change_by_selection(doc, selection, |range| {
-            let cursor = range.cursor(text);
-            let replacement_start = (cursor as i128 + start_offset) as usize;
-            let replacement_end = (cursor as i128 + end_offset) as usize;
-            let newline_with_offset = format!(
-                "{line_ending}{blank:width$}",
-                line_ending = line_ending,
-                width = replacement_start - doc.line_to_char(doc.char_to_line(replacement_start)),
-                blank = ""
-            );
-
-            let (replacement, tabstops) =
-                snippet::render(&snippet, newline_with_offset, include_placeholder);
-
-            let replacement_len = replacement.chars().count();
-            cursor_tabstop_offsets.push(
-                tabstops
-                    .first()
-                    .unwrap_or(&smallvec![(replacement_len, replacement_len)])
-                    .iter()
-                    .map(|(from, to)| -> (i128, i128) {
-                        (
-                            *from as i128 - replacement_len as i128,
-                            *to as i128 - replacement_len as i128,
-                        )
-                    })
-                    .collect(),
-            );
-
-            (replacement_start, replacement_end, Some(replacement.into()))
-        });
+        let (transaction, primary_idx) = Transaction::change_by_selection_ignore_overlapping(
+            doc,
+            selection,
+            |range, changeset| {
+                let cursor = range.cursor(text);
+                let (replacement_start, replacement_end) =
+                    completion_pos(text, start_offset, end_offset, cursor)?;
+                // Asssoc::After is used here because there should be no edits after
+                // `replacement_start` (those haven't been generated yet)
+                let mapped_replacement_start = changeset.map_pos(replacement_start, Assoc::After);
+                // if a snippet needs access to it's position (line number or offset) we must apply
+                // previous edits to the rope, as this has some ovherhead we do it lazily
+                // TODO: can we avoid apply the previous edits so this becomes O(N) instead of
+                // O(NlogN) with the number of selection ranges?
+                let mapped_doc = Lazy::new(|| {
+                    let mut mapped_doc = doc.clone();
+                    changeset.apply(&mut mapped_doc);
+                    mapped_doc
+                });
```

Comment:
This could probably be moved into change_by_selection_ignore_overlapping while still keeping it lazy (if the goal is to make Transaction object more powerful). Then this API can be updated to accept range and transaction_context, where transaction_context has a bunch of helper methods (e.g. getting current transformed doc, getting current changeset)

Reply:
> we try to keep the core transaction API as minimal as possible

This PR is already adding significant complexity to the transaction code. If you are optimizing for keeping the transaction code simpler, it may make sense to move all of this into the snippet rendering code (e.g. by creating a new transaction for each range and then just combining them).

Reply:
> > we try to keep the core transaction API as minimal as possible
> 
> This PR is already adding significant complexity to the transaction code and API. If you are optimizing for keeping the transaction code simpler, it may make sense to move all of this into the snippet rendering code (e.g. by creating a new transaction for each range and then just combining them).

especially with your suggestions above the API became quite minimal. Maybe minimal isn't even the right word. The API is supposed to be genric and not specical cased to one usecase. I don't want this general transaction logic inside the snippet/completion code either.

Reply:
I had similar concerns but the newest commits look a bit more straightforward 👍🏻